### PR TITLE
Implementing bypass deferred binding in Source Generator

### PR DESCRIPTION
**Pull request by surgupta-msft**
### Issue describing the changes in this PR

resolves #https://github.com/Azure/azure-functions-dotnet-worker/issues/1541

Changes are similar to commit - [FunctionMetadataGenerator.cs](https://github.com/Azure/azure-functions-dotnet-worker/commit/21002387335532201ff21781b5d66153b67dc413#diff-41d876a1aa0e264755d5373b41bbb7abed4ab5620d1a7fa60853c9ff6af8126c)

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: sdk/Sdk.Generators/Constants.cs [hunk 1]
```diff
@@ -34,6 +34,9 @@ internal static class Types
             internal const string HttpResponse = "Microsoft.Azure.Functions.Worker.Http.HttpResponseData";
             internal const string HttpTriggerBinding = "Microsoft.Azure.Functions.Worker.HttpTriggerAttribute";
 
+            internal const string InputConverterAttributeType = "Microsoft.Azure.Functions.Worker.Converters.InputConverterAttribute";
+            internal const string SupportedConverterTypeAttributeType = "Microsoft.Azure.Functions.Worker.Converters.SupportedConverterTypeAttribute";
+            internal const string SupportsDeferredBindingAttributeType = "Microsoft.Azure.Functions.Worker.Extensions.Abstractions.SupportsDeferredBindingAttribute";
```
Comment (satvu) [inline sdk/Sdk.Generators/Constants.cs:39]: You can also add these to [KnownFunctionMeatadataTypes](https://github.com/Azure/azure-functions-dotnet-worker/blob/main/sdk/Sdk.Generators/FunctionMetadataProviderGenerator/KnownFunctionMetadataTypes.cs) and load/store the symbols for comparison later during parsing.

File: sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs [hunk 2]
```diff
@@ -231,6 +240,75 @@ private bool TryGetParameterInputAndTriggerBindings(MethodDeclarationSyntax meth
...
+            }
+
+            private bool DoesConverterSupportDeferredBinding(TypedConstant converter, string bindingType)
+            {
+                var converterType = converter.Value as ITypeSymbol;
+                var converterAdvertisedAttributes = converterType?.GetAttributes().ToList();
+
+                if (converterAdvertisedAttributes is not null)
+                {
+                    bool converterAdvertisesDeferredBindingSupport = converterAdvertisedAttributes.Any(a => string.Equals(a.ToString(), Constants.Types.SupportsDeferredBindingAttributeType, StringComparison.Ordinal));
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs:273]: Could also use symbol comparison here

File: sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs [hunk 3]
```diff
@@ -137,7 +140,7 @@ private bool TryGetMethodOutputBinding(MethodDeclarationSyntax method, SemanticM
 
                 if (outputBindingAttribute != null)
                 {
-                    if (!TryCreateBindingDict(outputBindingAttribute, Constants.FunctionMetadataBindingProps.ReturnBindingName, bindingLocation, out IDictionary<string, object>? bindingDict))
+                    if (!TryCreateBindingDict(outputBindingAttribute, Constants.FunctionMetadataBindingProps.ReturnBindingName, bindingLocation, supportsDeferredBinding: false, out IDictionary<string, object>? bindingDict))
```
Comment (liliankasem) [inline sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs:143]: Can you make supportsDeferredBinding false by default in the `TryCreateBindingDict` method

**Review by liliankasem — APPROVED** [review]

**Review by satvu — APPROVED** [review]
lgtm!

File: sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs [hunk 4]
```diff
@@ -231,6 +240,75 @@ private bool TryGetParameterInputAndTriggerBindings(MethodDeclarationSyntax meth
...
 
+            private bool SupportsDeferredBinding(AttributeData bindingAttribute, string bindingType)
+            {
+                var advertisedAttributes = bindingAttribute?.AttributeClass?.GetAttributes();
+
+                if (advertisedAttributes != null)
+                {
+                    foreach (var advertisedAttribute in advertisedAttributes)
+                    {
+                        if (string.Equals(advertisedAttribute.AttributeClass?.GetFullName(), Constants.Types.InputConverterAttributeType, StringComparison.Ordinal))
```
[the remark this answers is not in the document]
Comment (surgupta-msft) [reply sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs]: Good suggestion

File: sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs [hunk 5]
```diff
@@ -231,6 +240,77 @@ private bool TryGetParameterInputAndTriggerBindings(MethodDeclarationSyntax meth
...
+                }
+
+                return false;
+            }
+
+            private bool DoesConverterSupportTargetType(List<AttributeData> converterAdvertisedAttributes, string bindingType)
+            {
+                foreach (AttributeData attribute in converterAdvertisedAttributes)
+                {
+                    if (string.Equals(attribute.AttributeClass?.GetFullName(), Constants.Types.SupportedConverterTypeAttributeType, StringComparison.Ordinal))
```
Comment (satvu) [inline sdk/Sdk.Generators/FunctionMetadataProviderGenerator/FunctionMetadataProviderGenerator.Parser.cs:298]: I think this section (and below) can also be refactored a bit.

**Pull request merged**